File: models/entity/game.go
package entity

import "time"

type Game struct {
	GameId                 int64     `xorm:"'id' pk autoincr" json:"id"`
	Title                  string    `xorm:"'title' nvarchar(64) notnull" json:"title"`                                     // 比赛标题
	Bio                    string    `xorm:"'bio' text" json:"bio"`                                                         // 比赛简介
	Description            string    `xorm:"'description' text" json:"description"`                                         // 比赛描述
	IsEnabled              *bool     `xorm:"'is_enabled' bool default(false) notnull" json:"is_enabled"`                    // 是否启用
	IsPublic               *bool     `xorm:"'is_public' default(true) notnull" json:"is_public"`                            // 是否为公开赛
	Password               string    `xorm:"'password' varchar(255)" json:"password"`                                       // 参赛密码，仅在内部赛中启用
	MemberLimitMin         int64     `xorm:"'member_limit_min' notnull default(1)" json:"member_limit_min"`                 // 团队人数最小值
	MemberLimitMax         int64     `xorm:"'member_limit_max' default(10)" json:"member_limit_max"`                        // 团队人数最大值
	ParallelContainerLimit int64     `xorm:"'parallel_container_limit' notnull default(2)" json:"parallel_container_limit"` // 容器并行限制
	FirstBloodRewardRatio  float64   `xorm:"'first_blood_reward_ratio' default(5)" json:"first_blood_reward_ratio"`         // 一血奖励比例
	SecondBloodRewardRatio float64   `xorm:"'second_blood_reward_ratio' default(3)" json:"second_blood_reward_ratio"`       // 二血奖励比例
	ThirdBloodRewardRatio  float64   `xorm:"'third_blood_reward_ratio' default(1)" json:"third_blood_reward_ratio"`         // 三血奖励比例
	IsNeedWriteUp          *bool     `xorm:"'is_need_write_up' bool default(true) notnull" json:"is_need_write_up"`         // 是否需要提交 WP
	StartedAt              time.Time `xorm:"'started_at' notnull" json:"started_at"`                                        // 比赛开始时间
	EndedAt                time.Time `xorm:"'ended_at' notnull" json:"ended_at"`                                            // 比赛结束时间
	CreatedAt              time.Time `xorm:"'created_at' created" json:"created_at"`                                        // 创建时间
	UpdatedAt              time.Time `xorm:"'updated_at' updated" json:"updated_at"`                                        // 更新时间
}

func (g *Game) TableName() string {
	return "game"
}

File: src/web/model/game.rs
use serde::{Deserialize, Serialize};
use validator::Validate;

use super::Metadata;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GetRequest {
    pub id: Option<i64>,
    pub title: Option<String>,
    pub is_enabled: Option<bool>,
    pub page: Option<u64>,
    pub size: Option<u64>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GetResponse {
    pub code: u16,
    pub data: Vec<crate::model::game::Model>,
    pub total: u64,
}

#[derive(Clone, Debug, Serialize, Deserialize, Validate)]
pub struct CreateRequest {
    pub title: String,
    pub started_at: i64,
    pub ended_at: i64,

    pub bio: Option<String>,
    pub description: Option<String>,
    pub is_enabled: Option<bool>,
    pub is_public: Option<bool>,
    pub member_limit_min: Option<i64>,
    pub member_limit_max: Option<i64>,
    pub parallel_container_limit: Option<i64>,
    pub is_need_write_up: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CreateResponse {
    pub code: u16,
    pub data: crate::model::game::Model,
}

#[derive(Clone, Debug, Serialize, Deserialize, Validate)]
pub struct UpdateRequest {
    pub id: Option<i64>,
    pub title: Option<String>,
    pub bio: Option<String>,
    pub description: Option<String>,
    pub is_enabled: Option<bool>,
    pub is_public: Option<bool>,
    pub member_limit_min: Option<i64>,
    pub member_limit_max: Option<i64>,
    pub parallel_container_limit: Option<i64>,
    pub is_need_write_up: Option<bool>,
    pub started_at: Option<i64>,
    pub ended_at: Option<i64>,
    pub frozed_at: Option<i64>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateResponse {
    pub code: u16,
    pub data: crate::model::game::Model,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DeleteResponse {
    pub code: u16,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GetChallengeRequest {
    pub game_id: Option<i64>,
    pub challenge_id: Option<i64>,
    pub team_id: Option<i64>,
    pub is_enabled: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GetChallengeResponse {
    pub code: u16,
    pub data: Vec<crate::model::game_challenge::Model>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CreateChallengeRequest {
    pub game_id: i64,
    pub challenge_id: i64,
    pub is_enabled: Option<bool>,
    pub difficulty: Option<i64>,
    pub max_pts: Option<i64>,
    pub min_pts: Option<i64>,
    pub first_blood_reward_ratio: Option<i64>,
    pub second_blood_reward_ratio: Option<i64>,
    pub third_blood_reward_ratio: Option<i64>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CreateChallengeResponse {
    pub code: u16,
    pub data: crate::model::game_challenge::Model,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateChallengeRequest {
    pub game_id: Option<i64>,
    pub challenge_id: Option<i64>,
    pub is_enabled: Option<bool>,
    pub difficulty: Option<i64>,
    pub max_pts: Option<i64>,
    pub min_pts: Option<i64>,
    pub first_blood_reward_ratio: Option<i64>,
    pub second_blood_reward_ratio: Option<i64>,
    pub third_blood_reward_ratio: Option<i64>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateChallengeResponse {
    pub code: u16,
    pub data: crate::model::game_challenge::Model,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DeleteChallengeResponse {
    pub code: u16,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GetTeamRequest {
    pub game_id: Option<i64>,
    pub team_id: Option<i64>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GetTeamResponse {
    pub code: u16,
    pub data: Vec<crate::model::game_team::Model>,
    pub total: u64,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CreateTeamRequest {
    pub game_id: i64,
    pub team_id: i64,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CreateTeamResponse {
    pub code: u16,
    pub data: crate::model::game_team::Model,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateTeamRequest {
    pub game_id: Option<i64>,
    pub team_id: Option<i64>,
    pub is_allowed: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateTeamResponse {
    pub code: u16,
    pub data: crate::model::game_team::Model,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DeleteTeamResponse {
    pub code: u16,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GetSubmissionRequest {
    pub status: Option<crate::model::submission::Status>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GetSubmissionResponse {
    pub code: u16,
    pub data: Vec<crate::model::submission::GameSubmissionModel>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GetPosterMetadataResponse {
    pub code: u16,
    pub data: Metadata,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SavePosterResponse {
    pub code: u16,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DeletePosterResponse {
    pub code: u16,
}
